src/protocol/boolean/prefix_or.rs

```diff
@@ -21,35 +21,47 @@ use super::BitOpStep;
 /// 5.2 Prefix-Or
 /// "Unconditionally Secure Constant-Rounds Multi-party Computation for Equality, Comparison, Bits, and Exponentiation"
 /// I. Damgård et al.
-pub struct PrefixOr<'a, B: BinaryField> {
-    input: &'a [Replicated<B>],
+pub struct PrefixOr<'a, F: Field> {
+    input: &'a [Replicated<F>],
 }
 
-impl<'a, B: BinaryField> PrefixOr<'a, B> {
+impl<'a, F: Field> PrefixOr<'a, F> {
     #[allow(dead_code)]
-    pub fn new(input: &'a [Replicated<B>]) -> Self {
+    pub fn new(input: &'a [Replicated<F>]) -> Self {
         Self { input }
     }
 
-    /// Securely computes `[a] | [b] where a, b ∈ {0, 1}`
-    /// OR can be computed as: `[a] ^ [b] ^ ([a] & [b])`
+    /// Securely computes `[a] | [b] where a, b ∈ {0, 1} ⊆ F_p`
+    ///
+    /// * OR can be computed as: `[a] ^ [b] ^ MULT([a], [b])`
+    /// * XOR([a], [b]) is: `[a] + [b] - 2([a] * [b])`
+    ///
+    /// Therefore,
+    ///
+    /// let [c] = [a] ^ [b]
+    /// [c] + [ab] - 2([c] * [ab])
     async fn bit_or(
-        a: Replicated<B>,
-        b: Replicated<B>,
-        ctx: ProtocolContext<'_, Replicated<B>, B>,
+        a: Replicated<F>,
+        b: Replicated<F>,
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
         record_id: RecordId,
-    ) -> Result<Replicated<B>, BoxError> {
-        let a_and_b = ctx.multiply(record_id, a, b).await?;
-        Ok(a + b + a_and_b)
+    ) -> Result<Replicated<F>, BoxError> {
+        let ab = ctx.narrow(&Step::AMultB).multiply(record_id, a, b).await?;
+        let c = a + b - (ab * F::from(2));
+        let cab = ctx
+            .narrow(&Step::ABMultC)
+            .multiply(record_id, c, ab)
+            .await?;
+        Ok(c + ab - (cab * F::from(2)))
     }
 
     /// Securely computes `∨ [a_1],...[a_n]`
     async fn block_or(
-        a: &[Replicated<B>],
+        a: &[Replicated<F>],
         k: usize,
-        ctx: ProtocolContext<'_, Replicated<B>, B>,
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
         record_id: RecordId,
-    ) -> Result<Replicated<B>, BoxError> {
+    ) -> Result<Replicated<F>, BoxError> {
         #[allow(clippy::cast_possible_truncation)]
         let mut v = a[0];
         for (i, &bit) in a[1..].iter().enumerate() {
```

Comment:
This is the part of the code that ought to use the "unbounded fan-in". This is a "symmetric function" because the output only depends on the number of 1s. In this case:

0 ones => 0
1 or more ones => 1

So we can compute the number of ones by just adding up all the secret shares of the bits. Great. Next, we need a function which has the following values:
1 => 0
2 => 1
3 => 1
4 => 1
5 => 1
6 => 1

(this is assuming the u32 case... it'll be slightly different otherwise).

Now we need to use Lagrange interpolation to find a polynomial that goes through all of these points. Once we have the coefficients of that function, we can just evaluate it.

Reply:
Per discussion offline, let me come back to this "symmetric function" part. I'll check and see which one (bit-by-bit or or an optimized symmetric function) performs better, but this optimization can be done later.